frontend/http/parser.go

```diff
@@ -128,7 +131,31 @@ func ParseScrape(r *http.Request, opts ParseOptions) (*bittorrent.ScrapeRequest,
 
 	infoHashes := qp.InfoHashes()
 	if len(infoHashes) < 1 {
-		return nil, bittorrent.ClientError("no info_hash parameter supplied")
+		// This is a potential fullscrape. Determine the origin IP and check if
+		// it's whitelisted.
+		host, _, err := net.SplitHostPort(r.RemoteAddr)
+		if err != nil {
+			return nil, err
+		}
+		remoteIP := net.ParseIP(host)
+		if remoteIP == nil {
+			return nil, errors.New("unable to parse HTTP remote address as IP")
```

Comment:
We parse IPs in a bunch of places. Maybe we should create one error for all of the failure cases across the codebase.

Reply:
Well, it's kinda messy. Sometimes we return the IP and an error, sometimes we return an IP and a bool... then there's also the whole IP spoofing thing. UDP only copies around bytes/remote socket address and does no checks whatsoever.

I don't think we can unify all of those :sweat_smile: